_source/_posts/2022-08-24-state-of-project-loom.md

```diff
@@ -0,0 +1,261 @@
+
+The new virtual threads in java 19 will be pretty easy to use. Compare the below with Golang's Goroutines or Kotlin's coroutines.
+
+**Virtual thread**
+
+```java
+Thread.startVirtualThread(() -> {
+    System.out.println("Hello, Project Loom!");
+});
+```
+
+**Goroutine**
+
+```go
+go func() {
+    println("Hello, Goroutines!")
+}()
+```
+
+**Kotlin coroutine**
+
+```kotlin
+runBlocking {
+    launch {
+        println("Hello, Kotlin coroutines!")
+    }
+}
+```
+
+A fun fact is that before Java 1.1, Java had support for green threads (a.k.a virtual threads). The new implementation of virtual threads is done in the JVM, where it maps multiple virtual threads to one or more OS threads, and the developer can use virtual threads or platform threads as per their needs. A few other important aspects of virtual threads are:
+
+- They are a `Thread` in code, runtime, debugger, and profiler
+- It's a Java entity and not a wrapper around a native thread
+- Creating and blocking them are cheap operations
+- They should not be pooled
+- It uses a work-stealing `ForkJoinPool` scheduler
+- Pluggable schedulers can be used for asynchronous programming
+- Virtual thread will have its own stack memory
+- API is very similar to OS threads and hence easier to adopt/migrate
+
+Let's look at some examples to show how powerful virtual threads are.
+
+#### Total number of threads
+
+First, let's see how many platform threads vs. virtual threads we can create on a machine (My machine is Intel Core i9-11900H, 8 cores, 16 threads, and 64GB RAM).
+
+**Platform threads**
+
+```java
+var counter = new AtomicInteger();
+while (true) {
+    new Thread(() -> {
+        int count = counter.incrementAndGet();
+        System.out.println("Thread count = " + count);
+        LockSupport.park();
+    }).start();
+}
+```
+
+On my machine, the code crashed after **32_539** platform threads.
+
+**Virtual threads**
+
+```java
+var counter = new AtomicInteger();
+while (true) {
+    Thread.startVirtualThread(() -> {
+        int count = counter.incrementAndGet();
+        System.out.println("Thread count = " + count);
+        LockSupport.park();
+    });
+}
+```
+
+On my machine, the process hanged after **14_625_956** virtual threads but didn't crash, and as memory became available, it kept going slowly.
+
+#### Task throughput
+
+Let's try to run 100,000 tasks using platform threads.
+
+```java
+try (var executor = Executors.newThreadPerTaskExecutor(Executors.defaultThreadFactory())) {
+    IntStream.range(0, 100_000).forEach(i -> executor.submit(() -> {
+        Thread.sleep(Duration.ofSeconds(1));
+        System.out.println(i);
+        return i;
+    }));
+}
+```
+
+This uses the `newThreadPerTaskExecutor` with the default thread factory and thus uses a thread group. If we run this code and time it, we get the below numbers. We get better performance if we use a thread pool with `Executors.newCachedThreadPool()`.
```

Comment:
This might read better as "I" vs "we". 

> When I ran this code and timed it, I get the below numbers. I get better performance if I use a thread pool with `Executors.newCachedThreadPool()`.